gpkit/constraints.py

```diff
@@ -0,0 +1,119 @@
+from abc import ABCMeta, abstractmethod, abstractproperty
+
+
+class Constraint(object):
+    __metaclass__ = ABCMeta
+
+    @abstractproperty
+    def varkeys(self):
+        "Varkeys present in the Constraint"
+        pass
+
+    @abstractproperty
+    def substitutions(self):
+        "Substitutions that the Constraint currently knows about"
+        pass
+    
+    @substitutions.setter
+    def substitutions(self, newvalue):
+        ".substitutions mist be settable"
+        pass
+
+    @abstractmethod
+    def __str__(self):
+        "string representation of the constraint, used for printing programs"
+        pass
+
+    @abstractmethod
+    def latex(self):
+        "latex representation of the constraint, used for printing programs"
+        pass
+
+    def _repr_latex_(self):
+        "Classes that inherit from Constraint will show as latex in IPython"
+        return "$$" + self.latex() + "$$"
+
+    @abstractmethod
+    def sub(self, subs, value=None):
+        "Returns new object without any of the varkeys in subs.keys()"
+        pass
+
+    @abstractmethod
+    def process_result(self, result):
+        """Does arbitrary computation / manipulation of a program's result
+
+        There's no guarantee what order different constraints will process results in,
+        so any changes made to the program's result should be careful not to step on
+        other constraints toes.
```

Comment:
Is it possible to say what the recommendation above means in practice? E.g. should we

 - advise against computed results that depend on anything other than the unprocessed `result`?
 - recommend that `process_result` never overwrite any field in `result`?
 - recommend that appending / adding a key is the only acceptable mutation of a `result`?